dango/testing/tests/dex.rs

```diff
@@ -466,3 +466,296 @@ fn submit_and_cancel_order_in_same_block() {
+    };
+    suite
+        .execute(
+            &mut accounts.owner,
+            contracts.dex,
+            &dex::ExecuteMsg::CreatePassivePool {
+                base_denom: DANGO_DENOM.clone(),
+                quote_denom: USDC_DENOM.clone(),
+                curve_type: CurveInvariant::Xyk,
+                lp_denom: lp_denom.clone(),
+                swap_fee: Udec128::ZERO,
+            },
+            initial_funds.clone(),
+        )
+        .should_succeed();
+
+    // provide liquidity
+    let provided_funds = coins! {
+        DANGO_DENOM.clone() => 100,
+        USDC_DENOM.clone() => 100,
+    };
+    suite
+        .execute(
+            &mut accounts.user1,
+            contracts.dex,
+            &dex::ExecuteMsg::ProvideLiquidity {
+                lp_denom: lp_denom.clone(),
+            },
+            provided_funds.clone(),
+        )
+        .should_succeed();
+
+    // record user and dex balances
+    suite
+        .balances()
+        .record_many(vec![accounts.user1.address(), contracts.dex.address()]);
+
+    // withdraw liquidity
+    suite
+        .execute(
+            &mut accounts.user1,
+            contracts.dex,
+            &dex::ExecuteMsg::WithdrawLiquidity {},
+            coins! {
+                lp_denom.clone() => withdraw_amount,
+            },
+        )
+        .should_succeed();
+
+    // Assert that the user's balances have changed as expected.
+    suite.balances().should_change(
+        accounts.user1.address(),
+        balance_changes_from_coins(
+            expected_funds_returned.clone(),
+            coins! { lp_denom.clone() => withdraw_amount },
+        ),
+    );
+
+    // Assert that the dex balance has decreased by the expected amount.
+    suite.balances().should_change(
+        contracts.dex.address(),
+        balance_changes_from_coins(Coins::new(), expected_funds_returned.clone()),
+    );
+
+    // Assert pool reserves are updated correctly
+    suite
+        .query_wasm_smart(contracts.dex, dango_types::dex::QueryPassivePoolRequest {
+            lp_denom: lp_denom.clone(),
+        })
+        .should_succeed_and(|pool| {
+            Coins::from(pool.reserves.clone())
+                == *initial_funds
+                    .clone()
+                    .insert_many(provided_funds)
+                    .unwrap()
+                    .deduct_many(expected_funds_returned)
+                    .unwrap()
+        });
+}
+
+fn balance_changes_from_coins(
+    increases: Coins,
+    decreases: Coins,
+) -> BTreeMap<Denom, BalanceChange> {
+    let mut changes: BTreeMap<Denom, BalanceChange> = increases
+        .into_iter()
+        .map(|Coin { denom, amount }| {
+            (denom.clone(), BalanceChange::Increased(amount.into_inner()))
+        })
+        .collect();
+    changes.extend(decreases.into_iter().map(|Coin { denom, amount }| {
+        (denom.clone(), BalanceChange::Decreased(amount.into_inner()))
+    }));
+    changes
+}
```

Comment:
should I put this in some place with shared test helpers? Or should I impl it on `BTreeMap<Denom, BalanceChange>`?